File: vscode-extension/src/lmcFormatter.ts
import * as vscode from 'vscode';

const LMC_INSTRUCTIONS = new Set([
    "INP", "OUT", "LDA", "STA", "ADD", "SUB", "BRA", "BRZ", "BRP", "HLT", "DAT"
]);

// Pattern to match a line: [LABEL] INSTRUCTION [OPERAND] [// COMMENT]
const LMC_LINE_PATTERN = new RegExp(
    "^\\s*" +                               // Optional leading whitespace
    "(?:([A-Za-z_][A-Za-z0-9_]*)\\s*:\s*)?" + // Optional Label (Group 1) with optional colon and whitespace
    "([A-Za-z]{3})" +                      // Instruction (Group 2)
    "(?:\\s+([A-Za-z_][A-Za-z0-9_]*|\\d+))?" + // Optional Operand (Group 3: label or number)
    "(?:\\s*//.*)?" +                     // Optional Comment
    "\\s*$"                                // Optional trailing whitespace and end of line
);

export function formatLmcDocument(document: vscode.TextDocument): vscode.TextEdit[] {
    const edits: vscode.TextEdit[] = [];
    const lines = document.getText().split(/\r?\n/);

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        const trimmedLine = line.trim();

        if (trimmedLine.length === 0) {
            continue; // Skip empty lines
        }

        // Preserve full-line comments
        if (trimmedLine.startsWith("//") || trimmedLine.startsWith(";")) {
            // Ensure comments are not indented if they are full-line comments
            if (line !== trimmedLine) {
                edits.push(vscode.TextEdit.replace(
                    new vscode.Range(i, 0, i, line.length),
                    trimmedLine
                ));
            }
            continue;
        }

        const match = trimmedLine.match(LMC_LINE_PATTERN);
        if (match) {
            let formattedLine = "";
            const label = match[1];
            const instruction = match[2];
            const operand = match[3];
            const commentIndex = trimmedLine.indexOf("//");
            const comment = commentIndex !== -1 ? trimmedLine.substring(commentIndex) : null;

            // Auto capitalize label and append
            if (label) {
                formattedLine += label.toUpperCase() + ":\t";
            }

            // Auto capitalize instruction and append
            formattedLine += instruction.toUpperCase();

            // Append operand if present
            if (operand) {
                formattedLine += "\t" + operand.toUpperCase();
            }

            // Append comment if present
            if (comment) {
                // Align comments if there's an instruction/operand
                if (label || instruction || operand) {
                    formattedLine += "\t"; // Add a tab for alignment
                }
                formattedLine += comment;
            }

            // Apply basic indentation for lines without labels
            if (!label && formattedLine.length > 0 && !formattedLine.startsWith("//") && !formattedLine.startsWith(";")) {
                formattedLine = "\t" + formattedLine; // Indent with one tab
            }

            if (formattedLine !== line) {
                edits.push(vscode.TextEdit.replace(
                    new vscode.Range(i, 0, i, line.length),
                    formattedLine
                ));
            }
        } else {
            // If line doesn't match pattern, just ensure it's trimmed
            if (line !== trimmedLine) {
                edits.push(vscode.TextEdit.replace(
                    new vscode.Range(i, 0, i, line.length),
                    trimmedLine
                ));
            }
        }
    }

    return edits;
}

File: ide-javafx/src/main/java/com/lmc/ide/LMCFormatter.java
package com.lmc.ide;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LMCFormatter {
    private static final Pattern LINE_PATTERN = Pattern.compile(
            "^\\s*" +
                    "(?:([A-Za-z_][A-Za-z0-9_]*):\\s*)?" +
                    "(INP|OUT|LDA|STA|ADD|SUB|BRA|BRZ|BRP|HLT|DAT)" +
                    "(?:\\s+([A-Za-z_][A-Za-z0-9_]*|\\d+))?" +
                    "(\\s*//.*)?" +
                    "\\s*$",
            Pattern.CASE_INSENSITIVE);

    public static String format(String lmcCode) {
        StringBuilder formattedCode = new StringBuilder();
        String[] lines = lmcCode.split("\\r?\\n");
        for (String line : lines) {
            String trimmedLine = line.trim();
            if (trimmedLine.isEmpty() || trimmedLine.startsWith("//") || trimmedLine.startsWith("#")) {
                formattedCode.append(trimmedLine).append("\n");
                continue;
            }
            Matcher matcher = LINE_PATTERN.matcher(trimmedLine);
            if (matcher.matches()) {
                String label = matcher.group(1);
                String instruction = matcher.group(2);
                String operand = matcher.group(3);
                String comment = matcher.group(4);
                StringBuilder currentLine = new StringBuilder();
                if (label != null) {
                    currentLine.append(String.format("%-8s", label.toUpperCase() + ":"));
                } else {
                    currentLine.append("\t");
                }
                currentLine.append(instruction.toUpperCase());
                if (operand != null) {
                    currentLine.append("\t").append(operand.toUpperCase());
                }
                if (comment != null) {
                    currentLine.append("\t").append(comment.trim());
                }
                formattedCode.append(currentLine.toString()).append("\n");
            } else {
                formattedCode.append(trimmedLine).append("\n");
            }
        }
        return formattedCode.toString();
    }
}
